JIT get_reduced_triangular

ulupo opened the pull request:
When added to previous jitting, 10x speedup in the simple example used so far is achieved.  But I had to give up the possibility that `matrix` be a tuple (or else it won't compile, at least not in this form).

ulupo committed:
Fix list comp

ulupo committed:
njit get_reduced_triangular

ulupo commented on steenroder/__init__.py:
```diff
@@ -1,19 +1,19 @@
 import numpy as np
 from itertools import combinations
-from numba import jit
+from numba import njit
```
`njit` is short-hand for `jit(nopython=True)`

ulupo commented on steenroder/__init__.py:
```diff
@@ -31,19 +31,21 @@ def get_coboundary(filtration):
-    reduced = np.array(matrix)
-    triangular = np.eye(n, dtype=np.bool)
+    reduced = matrix.copy()
+    triangular = np.zeros((n, n), dtype=np.bool_)
```
Inside `numba`, "real" numpy dtypes must be used instead of python in-built types such as `bool`, `float`, etc.  Note that `np.bool` is just an alias for `bool`, it is `np.bool_` which is the *real* numpy dtype.

ulupo commented on steenroder/__init__.py:
```diff
@@ -1,19 +1,19 @@
 def get_boundary(filtration):
     spx_filtration_idx = {tuple(v): idx for idx, v in enumerate(filtration)}
-    boundary = np.zeros((len(filtration), len(filtration)), dtype=np.bool)
+    boundary = np.zeros((len(filtration), len(filtration)), dtype=bool)
```
New versions of numpy state that `dtype=np.bool` is deprecated and one should just pass the in-built Python `bool`.  Same for `np.float` etc.  Of course, when using `numba` neither `np.bool` nor `bool` is correct, see below.

ulupo commented on steenroder/__init__.py:
```diff
@@ -234,7 +236,9 @@ def get_steenrod_barcode(reduced, steenrod_matrix):
                 if j > i:
                     barcode.append((i, j))
-    barcode += [(i, np.inf) for i in alive if alive[i]]
+    for i in alive:
```
List comprehensions are not allowed in `numba`.

ulupo replied:
Actually this may not be true. I'll investigate

ammedmar commented:
I leave this [link](https://drive.google.com/drive/folders/1ieBaMqDdDDXrChaAH8F_ja6Wlb_KuPWl?usp=sharing) here in case it is useful. It contains bigger examples.

ulupo committed:
Linting

ulupo committed:
Small optimizations, make STSQ return a cochain

ulupo committed:
Speed-up by avoiding np.concatenate

ulupo committed:
Faster pivot function

ulupo committed:
Speed-up by using Fortran order

Slicing columns is faster this way.  Numba does not support the order kwarg in np.zeros and np.empty, so transposition was used instead.

ulupo committed:
Make specialized faster function for checking that two pivots are the same

ulupo committed:
Revert to symmetric differences following @ammedmar's review
